tox.ini

```diff
@@ -6,8 +6,8 @@ allowlist_externals = pytest
 setenv=
   PYTHONHASHSEED=0
 deps=
-  -rdev-requirements.txt
-  .
+  -crequirements.txt
```

Comment:
Why does this move to a constraints (-c) for versions instead of requirements (-r). I feel like this has broke my dev workflow when it was done in a different project when trying to install from the command line and include the `dev` extras.

Reply:
1. Remove dev-requirements.txt remove pins for for develop dependencies. We only need pins for production deps. This remove some extra work for developers. So we only have to use pip-tools to compile requirements.txt, and do not need to compile dev-dependencies anymore.
2. we could use -r requirements.txt here. I used -c requirements.txt here because I think this is faster. If we use -r requirements.txt, pip has to calculate all the dependencies. By using -c requirements.txt, pip will use them as constrain, and only calculate ".[dev]" dependencies, which suppose to be faster.
3. When using the requimrents.txt file as constrain, the requirements.txt file can not have extra deps for packages. one example I have saw is something like pyjwt[crypto]==2.8.0, this will not work if it exists in requirements.txt.

Reply:
I'm not sure if the internet is giving me good information, but I believe requirements and constraints can be used  conjointly like -r requirements.txt -c constraints.txt but they do have different behaviors.

So I guess I have couple of questions:
1. if you try and install the project with a clean python installation will it download and install all the modules you need if only use `-c`? 
2. do we have a reproducible build if we specify constraints but let 'pip install' pull non constrained modules?
3. Is it more appropriate to have a constraints.txt than requirements.txt? Should we be using both?

All of this could be ignorance on my part, but I've seen conflicting information and 'pip install --help' does not give me the advice I need.

Reply:
1. if you try and install the project with a clean python installation will it download and install all the modules you need if only use -c?
The full command is `pip install -c requirements.txt .[dev]`.
`pip install .[dev]` will install all the modules we need for testing (this is tox config file). 
`-c requirements.txt` make sure we are testing the same dependencies we are using in produciton.

2. do we have a reproducible build if we specify constraints but let 'pip install' pull non constrained modules?
I think for reproducible build, we are building for production. 
This is a tox config file and is for testing. And yes, I use the requirements.txt as a constraint file to make sure we are testing the same dependencies in production. 

3. Is it more appropriate to have a constraints.txt than requirements.txt? Should we be using both?
I do not think we should use both. It should be a single source of truth. 
This requirements.txt is also used to install packages in production, it is really a requirements file then. 
It is only used as a constraints file when installing for testing. 
So I think we should keep the name.

Reply:
Hopefully final question, do you think it is wise to build for development differently than you do for production?

Reply:
Build for development has always been different from production, because we need extra packages for development. Like pytest, indexd server, etc.

But, all the packages we are using in production, we use the same package when we do testing. That is what the `-c requriements.txt` is for. Which has the same end result as using `-r requirements.txt`. Except use the constraints is faster.